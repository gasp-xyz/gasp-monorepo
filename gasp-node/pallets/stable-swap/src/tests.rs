use std::vec;

use frame_support::{assert_err, assert_ok};
use sp_runtime::BoundedVec;

use crate::{assert_event_emitted, mock::*, Config, Error, Event, Pallet};

const UNIT: u128 = 10_u128.pow(18);

fn prep_pool() -> (AccountId, Balance, Balance) {
	let account: AccountId = 2;
	let amount: Balance = 1_000_000 * UNIT;
	let mint: Balance = 1_000 * UNIT;

	StableSwap::create_new_token(&account, amount);
	StableSwap::create_new_token(&account, amount);
	StableSwap::create_new_token(&account, amount);
	StableSwap::create_pool(
		RuntimeOrigin::signed(account),
		vec![0, 1, 2],
		vec![UNIT, UNIT, UNIT],
		200,
	)
	.unwrap();

	assert_ok!(StableSwap::add_liquidity(
		RuntimeOrigin::signed(account),
		3,
		vec![mint, mint, mint],
		0,
	));

	return (account, amount, mint)
}

// create pool tests

#[test]
fn create_pool_should_work() {
	new_test_ext().execute_with(|| {
		let account: AccountId = 2;
		let amount: Balance = 1_000_000_000;
		let rate: u128 = 10_u128.pow(18);

		StableSwap::create_new_token(&account, amount);
		StableSwap::create_new_token(&account, amount);
		StableSwap::create_new_token(&account, amount);

		StableSwap::create_pool(
			RuntimeOrigin::signed(account),
			vec![0, 1, 2],
			vec![rate, rate, rate],
			100,
		)
		.unwrap();

		assert!(<Test as Config>::Currency::exists(3));
	});
}

#[test]
fn create_pool_should_fail_on_nonexistent_asset() {
	new_test_ext().execute_with(|| {
		let account: AccountId = 2;
		let amount: Balance = 1_000_000_000;
		let rate: u128 = 10_u128.pow(18);

		StableSwap::create_new_token(&account, amount);
		StableSwap::create_new_token(&account, amount);

		assert_err!(
			StableSwap::create_pool(
				RuntimeOrigin::signed(account),
				vec![0, 1, 2],
				vec![rate, rate, rate],
				100
			),
			Error::<Test>::AssetDoesNotExist,
		);
	});
}

#[test]
fn create_pool_should_fail_on_same_asset() {
	new_test_ext().execute_with(|| {
		let account: AccountId = 2;
		let amount: Balance = 1_000_000_000;
		let rate: u128 = 10_u128.pow(18);

		StableSwap::create_new_token(&account, amount);
		StableSwap::create_new_token(&account, amount);

		assert_err!(
			StableSwap::create_pool(
				RuntimeOrigin::signed(account),
				vec![0, 1, 1],
				vec![rate, rate, rate],
				100
			),
			Error::<Test>::SameAsset,
		);
	});
}

#[test]
fn create_pool_should_fail_on_too_many_assets() {
	new_test_ext().execute_with(|| {
		let account: AccountId = 2;
		let amount: Balance = 1_000_000_000;
		let rate: u128 = 10_u128.pow(18);

		StableSwap::create_new_token(&account, amount);
		StableSwap::create_new_token(&account, amount);

		assert_err!(
			StableSwap::create_pool(
				RuntimeOrigin::signed(account),
				vec![0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
				vec![rate, rate, rate],
				100
			),
			Error::<Test>::TooManyAssets,
		);
	});
}

#[test]
fn create_pool_should_fail_on_coeff_out_out_range() {
	new_test_ext().execute_with(|| {
		let account: AccountId = 2;
		let amount: Balance = 1_000_000_000;
		let rate: u128 = 10_u128.pow(18);

		StableSwap::create_new_token(&account, amount);
		StableSwap::create_new_token(&account, amount);

		assert_err!(
			StableSwap::create_pool(
				RuntimeOrigin::signed(account),
				vec![0, 1],
				vec![rate, rate, rate],
				0
			),
			Error::<Test>::AmpCoeffOutOfRange,
		);
		assert_err!(
			StableSwap::create_pool(
				RuntimeOrigin::signed(account),
				vec![0, 1],
				vec![rate, rate, rate],
				u128::MAX
			),
			Error::<Test>::AmpCoeffOutOfRange,
		);
	});
}

// add liquidity tests
#[test]
fn add_liquidity_should_work() {
	new_test_ext().execute_with(|| {
		let (account, _, mint) = prep_pool();

		// check first balanced add
		assert_event_emitted!(Event::LiquidityMinted {
			who: 2,
			pool_id: 3,
			amounts_provided: BoundedVec::truncate_from(vec![mint, mint, mint]),
			lp_token: 3,
			lp_token_minted: 3_000 * UNIT,
			total_supply: 3_000 * UNIT,
			fees: BoundedVec::truncate_from(vec![]),
		});

		assert_eq!(StableSwap::get_virtual_price(&3).unwrap(), 1 * UNIT);

		let amounts = vec![5 * mint, mint, 40 * mint];
		let expected = StableSwap::calc_lp_token_amount(&3, amounts.clone(), true).unwrap();

		// imbalanced add, should have fees
		assert_ok!(StableSwap::add_liquidity(
			RuntimeOrigin::signed(account),
			3,
			amounts.clone(),
			1,
		));

		assert_event_emitted!(Event::LiquidityMinted {
			who: 2,
			pool_id: 3,
			amounts_provided: BoundedVec::truncate_from(amounts),
			lp_token: 3,
			lp_token_minted: expected,
			total_supply: 3_000 * UNIT + 45322880144288283584179,
			fees: BoundedVec::truncate_from(vec![
				12498711677479364060,
				21094110011003080268,
				30692298654602638774
			]),
		});

		// half of fees goes to treasury
		assert_eq!(StableSwap::balance(0, TreasuryAccount::get()), 4166237226659702131);
		assert_eq!(StableSwap::balance(1, TreasuryAccount::get()), 7031370005073967423);
		assert_eq!(StableSwap::balance(2, TreasuryAccount::get()), 10230766220247032834);
		assert_eq!(StableSwap::get_virtual_price(&3).unwrap(), 1000961844675256200);
	});
}

#[test]
fn add_liquidity_balanced_for_single_asset_minimal_fees() {
	new_test_ext().execute_with(|| {
		let account: AccountId = 2;
		let amount: Balance = 1_000_000 * UNIT;

		StableSwap::create_new_token(&account, amount);
		StableSwap::create_new_token(&account, amount);
		StableSwap::create_new_token(&account, amount);
		StableSwap::create_pool(
			RuntimeOrigin::signed(account),
			vec![0, 1, 2],
			vec![UNIT, UNIT, UNIT],
			200,
		)
		.unwrap();

		assert_ok!(StableSwap::add_liquidity(
			RuntimeOrigin::signed(account),
			3,
			vec![2 * UNIT, 3 * UNIT, 10 * UNIT],
			1,
		));

		let input_0 = 10_000 * UNIT;
		let reserves = StableSwap::get_pool_reserves(&3).unwrap();
		let rate = input_0 / reserves[0];
		let amounts = vec![input_0, reserves[1] * rate, reserves[2] * rate];
		let exp = Pallet::<Test>::calc_lp_token_amount(&3, amounts.clone(), true).unwrap();

		assert_ok!(StableSwap::add_liquidity(
			RuntimeOrigin::signed(account),
			3,
			amounts.clone(),
			1,
		));

		assert_event_emitted!(Event::LiquidityMinted {
			who: 2,
			pool_id: 3,
			amounts_provided: BoundedVec::truncate_from(amounts),
			lp_token: 3,
			lp_token_minted: exp,
			total_supply: 74881186777958934408403,
			fees: BoundedVec::truncate_from(vec![0, 0, 1]),
		});
	});
}

// remove liquidity
#[test]
fn remove_liquidity_one_asset_should_work() {
	new_test_ext().execute_with(|| {
		let (account, _, _) = prep_pool();
		let total_supply = StableSwap::total_supply(3);

		assert_ok!(StableSwap::remove_liquidity_one_asset(
			RuntimeOrigin::signed(account),
			3,
			0,
			10 * UNIT,
			0
		));

		assert_event_emitted!(Event::LiquidityBurnedOne {
			who: 2,
			pool_id: 3,
			asset_id: 0,
			// a bit less then 10, minus fees
			amount: 9_984_833_787_927_106_037,
			burned_amount: 10 * UNIT,
			total_supply: total_supply - 10 * UNIT,
		});

		assert_eq!(StableSwap::balance(0, TreasuryAccount::get()), 4999816321852351);
		assert_eq!(StableSwap::balance(1, TreasuryAccount::get()), 0);
		assert_eq!(StableSwap::balance(2, TreasuryAccount::get()), 0);
	});
}

#[test]
fn remove_liquidity_imbalanced_should_work() {
	new_test_ext().execute_with(|| {
		let (account, balance, mint) = prep_pool();
		let amounts = vec![mint / 100, mint / 50, mint / 200];

		assert_eq!(StableSwap::balance(0, 2), balance - mint);
		assert_eq!(StableSwap::balance(1, 2), balance - mint);
		assert_eq!(StableSwap::balance(2, 2), balance - mint);

		assert_ok!(StableSwap::remove_liquidity_imbalanced(
			RuntimeOrigin::signed(account),
			3,
			amounts.clone(),
			u128::MAX,
		));

		let total_supply = StableSwap::total_supply(3);
		assert_event_emitted!(Event::LiquidityBurned {
			who: 2,
			pool_id: 3,
			amounts: BoundedVec::truncate_from(amounts.clone()),
			burned_amount: 35019044399900264325,
			total_supply,
			fees: BoundedVec::truncate_from(vec![
				1875110298930542,
				9374909700834153,
				7500120299077607,
			]),
		});

		// got what requested
		assert_eq!(StableSwap::balance(0, 2), balance - mint + amounts[0]);
		assert_eq!(StableSwap::balance(1, 2), balance - mint + amounts[1]);
		assert_eq!(StableSwap::balance(2, 2), balance - mint + amounts[2]);
		assert_eq!(StableSwap::balance(3, 2), 3 * mint - 35019044399900264325);

		assert_eq!(StableSwap::balance(0, TreasuryAccount::get()), 625036766435188);
		assert_eq!(StableSwap::balance(1, TreasuryAccount::get()), 3124969900903044);
		assert_eq!(StableSwap::balance(2, TreasuryAccount::get()), 2500040100192543);
	});
}

#[test]
fn remove_liquidity_to_zero_should_work() {
	new_test_ext().execute_with(|| {
		let (account, balance, mint) = prep_pool();
		let total_supply = StableSwap::total_supply(3);

		assert_ok!(StableSwap::remove_liquidity(
			RuntimeOrigin::signed(account),
			3,
			total_supply,
			vec![mint, mint, mint],
		));

		assert_event_emitted!(Event::LiquidityBurned {
			who: 2,
			pool_id: 3,
			amounts: BoundedVec::truncate_from(vec![mint, mint, mint]),
			burned_amount: 3 * mint,
			total_supply: 0,
			fees: BoundedVec::truncate_from(vec![]),
		});

		// all back
		assert_eq!(StableSwap::balance(0, 2), balance);
		assert_eq!(StableSwap::balance(1, 2), balance);
		assert_eq!(StableSwap::balance(2, 2), balance);
		assert_eq!(StableSwap::balance(3, 2), 0);

		assert_eq!(StableSwap::balance(0, TreasuryAccount::get()), 0);
		assert_eq!(StableSwap::balance(1, TreasuryAccount::get()), 0);
		assert_eq!(StableSwap::balance(2, TreasuryAccount::get()), 0);
	});
}

// swaps
#[test]
fn swap_should_work_dy() {
	new_test_ext().execute_with(|| {
		let (account, _, _) = prep_pool();

		let dy = StableSwap::get_dy(&3, 0, 2, 100 * UNIT).unwrap();

		assert_ok!(StableSwap::swap(RuntimeOrigin::signed(account), 3, 0, 2, 100 * UNIT, 0));

		assert_event_emitted!(Event::AssetsSwapped {
			who: 2,
			pool_id: 3,
			asset_in: 0,
			amount_in: 100 * UNIT,
			asset_out: 2,
			amount_out: dy
		});

		assert_eq!(StableSwap::balance(0, TreasuryAccount::get()), 0);
		assert_eq!(StableSwap::balance(1, TreasuryAccount::get()), 0);
		assert_eq!(StableSwap::balance(2, TreasuryAccount::get()), 50037401982926047);
	});
}

#[test]
fn swap_should_work_dx() {
	new_test_ext().execute_with(|| {
		let (account, _, _) = prep_pool();

		let dx = StableSwap::get_dx(&3, 0, 2, 100 * UNIT).unwrap();

		assert_ok!(StableSwap::swap(RuntimeOrigin::signed(account), 3, 0, 2, dx, 0));

		assert_event_emitted!(Event::AssetsSwapped {
			who: 2,
			pool_id: 3,
			asset_in: 0,
			amount_in: dx,
			asset_out: 2,
			amount_out: 100 * UNIT,
		});

		assert_eq!(StableSwap::balance(0, TreasuryAccount::get()), 0);
		assert_eq!(StableSwap::balance(1, TreasuryAccount::get()), 0);
		assert_eq!(StableSwap::balance(2, TreasuryAccount::get()), 50213816221710612);
	});
}
