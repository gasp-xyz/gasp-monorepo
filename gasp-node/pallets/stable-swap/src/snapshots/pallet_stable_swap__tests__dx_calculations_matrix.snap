---
source: pallets/stable-swap/src/tests.rs
assertion_line: 492
expression: results
snapshot_kind: text
---
- amp_coeff: 200
  initial_liquidity:
    - 500000000000000000000
    - 10000000000000000000000
  rate_multipliers:
    - 1000000000000000000
    - 1000000000000000000
  dy: 100000000000000000
  calculated_dx: 79621311568089795
- amp_coeff: 1000
  initial_liquidity:
    - 500000000000000000000
    - 10000000000000000000000
  rate_multipliers:
    - 1000000000000000000
    - 1000000000000000000
  dy: 100000000000000000
  calculated_dx: 95320109027122834
- amp_coeff: 1000
  initial_liquidity:
    - 1000000000000000000000
    - 1000000000000000000000
  rate_multipliers:
    - 1000000000000000000
    - 1000000000000000000
  dy: 100000000000000000
  calculated_dx: 100300912758346345
- amp_coeff: 200
  initial_liquidity:
    - 1000000000000000000000
    - 1000000000000000000000
  rate_multipliers:
    - 1000000000000000000
    - 1000000000000000000
  dy: 100000000000000000
  calculated_dx: 100300952759249769
- amp_coeff: 200
  initial_liquidity:
    - 500000000000000000000
    - 10000000000000000000000
  rate_multipliers:
    - 1000000000000000000
    - 2000000000000000000
  dy: 100000000000000000
  calculated_dx: 102732403960421427
- amp_coeff: 1000
  initial_liquidity:
    - 500000000000000000000
    - 10000000000000000000000
  rate_multipliers:
    - 1000000000000000000
    - 2000000000000000000
  dy: 100000000000000000
  calculated_dx: 166765400466768106
- amp_coeff: 200
  initial_liquidity:
    - 1000000000000000000000
    - 1000000000000000000000
  rate_multipliers:
    - 1000000000000000000
    - 2000000000000000000
  dy: 100000000000000000
  calculated_dx: 199798972640443209
- amp_coeff: 1000
  initial_liquidity:
    - 1000000000000000000000
    - 1000000000000000000000
  rate_multipliers:
    - 1000000000000000000
    - 2000000000000000000
  dy: 100000000000000000
  calculated_dx: 200468383268724300
- amp_coeff: 200
  initial_liquidity:
    - 500000000000000000000
    - 10000000000000000000000
  rate_multipliers:
    - 1000000000000000000
    - 1000000000000000000
  dy: 10000000000000000000
  calculated_dx: 7987683472135033225
- amp_coeff: 1000
  initial_liquidity:
    - 500000000000000000000
    - 10000000000000000000000
  rate_multipliers:
    - 1000000000000000000
    - 1000000000000000000
  dy: 10000000000000000000
  calculated_dx: 9541078391370206537
- amp_coeff: 1000
  initial_liquidity:
    - 1000000000000000000000
    - 1000000000000000000000
  rate_multipliers:
    - 1000000000000000000
    - 1000000000000000000
  dy: 10000000000000000000
  calculated_dx: 10030191163399874055
- amp_coeff: 200
  initial_liquidity:
    - 1000000000000000000000
    - 1000000000000000000000
  rate_multipliers:
    - 1000000000000000000
    - 1000000000000000000
  dy: 10000000000000000000
  calculated_dx: 10030591235964353940
- amp_coeff: 200
  initial_liquidity:
    - 500000000000000000000
    - 10000000000000000000000
  rate_multipliers:
    - 1000000000000000000
    - 2000000000000000000
  dy: 10000000000000000000
  calculated_dx: 10373084326856906302
- amp_coeff: 1000
  initial_liquidity:
    - 500000000000000000000
    - 10000000000000000000000
  rate_multipliers:
    - 1000000000000000000
    - 2000000000000000000
  dy: 10000000000000000000
  calculated_dx: 16767999077645334338
- amp_coeff: 200
  initial_liquidity:
    - 1000000000000000000000
    - 1000000000000000000000
  rate_multipliers:
    - 1000000000000000000
    - 2000000000000000000
  dy: 10000000000000000000
  calculated_dx: 19982188323218663237
- amp_coeff: 1000
  initial_liquidity:
    - 1000000000000000000000
    - 1000000000000000000000
  rate_multipliers:
    - 1000000000000000000
    - 2000000000000000000
  dy: 10000000000000000000
  calculated_dx: 20047185309026867197
- amp_coeff: 200
  initial_liquidity:
    - 500000000000000000000
    - 10000000000000000000000
  rate_multipliers:
    - 1000000000000000000
    - 1000000000000000000
  dy: 50000000000000000000
  calculated_dx: 40431405241148586002
- amp_coeff: 1000
  initial_liquidity:
    - 500000000000000000000
    - 10000000000000000000000
  rate_multipliers:
    - 1000000000000000000
    - 1000000000000000000
  dy: 50000000000000000000
  calculated_dx: 47874030321230232735
- amp_coeff: 1000
  initial_liquidity:
    - 1000000000000000000000
    - 1000000000000000000000
  rate_multipliers:
    - 1000000000000000000
    - 1000000000000000000
  dy: 50000000000000000000
  calculated_dx: 50153017817720733956
- amp_coeff: 200
  initial_liquidity:
    - 1000000000000000000000
    - 1000000000000000000000
  rate_multipliers:
    - 1000000000000000000
    - 1000000000000000000
  dy: 50000000000000000000
  calculated_dx: 50163045998991602835
- amp_coeff: 200
  initial_liquidity:
    - 500000000000000000000
    - 10000000000000000000000
  rate_multipliers:
    - 1000000000000000000
    - 2000000000000000000
  dy: 50000000000000000000
  calculated_dx: 53871982583804960915
- amp_coeff: 200
  initial_liquidity:
    - 500000000000000000000
    - 10000000000000000000000
  rate_multipliers:
    - 1000000000000000000
    - 1000000000000000000
  dy: 100000000000000000000
  calculated_dx: 81996143870959046344
- amp_coeff: 1000
  initial_liquidity:
    - 500000000000000000000
    - 10000000000000000000000
  rate_multipliers:
    - 1000000000000000000
    - 2000000000000000000
  dy: 50000000000000000000
  calculated_dx: 85499283666606430685
- amp_coeff: 1000
  initial_liquidity:
    - 500000000000000000000
    - 10000000000000000000000
  rate_multipliers:
    - 1000000000000000000
    - 1000000000000000000
  dy: 100000000000000000000
  calculated_dx: 96113260059407330002
- amp_coeff: 200
  initial_liquidity:
    - 1000000000000000000000
    - 1000000000000000000000
  rate_multipliers:
    - 1000000000000000000
    - 2000000000000000000
  dy: 50000000000000000000
  calculated_dx: 99953801549151109890
- amp_coeff: 1000
  initial_liquidity:
    - 1000000000000000000000
    - 1000000000000000000000
  rate_multipliers:
    - 1000000000000000000
    - 2000000000000000000
  dy: 50000000000000000000
  calculated_dx: 100242375410966136220
- amp_coeff: 1000
  initial_liquidity:
    - 1000000000000000000000
    - 1000000000000000000000
  rate_multipliers:
    - 1000000000000000000
    - 1000000000000000000
  dy: 100000000000000000000
  calculated_dx: 100311436021492352809
- amp_coeff: 200
  initial_liquidity:
    - 1000000000000000000000
    - 1000000000000000000000
  rate_multipliers:
    - 1000000000000000000
    - 1000000000000000000
  dy: 100000000000000000000
  calculated_dx: 100351863655001096377
- amp_coeff: 200
  initial_liquidity:
    - 500000000000000000000
    - 10000000000000000000000
  rate_multipliers:
    - 1000000000000000000
    - 2000000000000000000
  dy: 100000000000000000000
  calculated_dx: 112670097631551686719
- amp_coeff: 200
  initial_liquidity:
    - 500000000000000000000
    - 10000000000000000000000
  rate_multipliers:
    - 1000000000000000000
    - 1000000000000000000
  dy: 200000000000000000000
  calculated_dx: 167937292627340845773
- amp_coeff: 1000
  initial_liquidity:
    - 500000000000000000000
    - 10000000000000000000000
  rate_multipliers:
    - 1000000000000000000
    - 2000000000000000000
  dy: 100000000000000000000
  calculated_dx: 174423333632934110526
- amp_coeff: 1000
  initial_liquidity:
    - 500000000000000000000
    - 10000000000000000000000
  rate_multipliers:
    - 1000000000000000000
    - 1000000000000000000
  dy: 200000000000000000000
  calculated_dx: 193400270505847093835
- amp_coeff: 200
  initial_liquidity:
    - 1000000000000000000000
    - 1000000000000000000000
  rate_multipliers:
    - 1000000000000000000
    - 2000000000000000000
  dy: 100000000000000000000
  calculated_dx: 200002321868636107481
- amp_coeff: 1000
  initial_liquidity:
    - 1000000000000000000000
    - 1000000000000000000000
  rate_multipliers:
    - 1000000000000000000
    - 2000000000000000000
  dy: 100000000000000000000
  calculated_dx: 200498945273506619787
- amp_coeff: 1000
  initial_liquidity:
    - 1000000000000000000000
    - 1000000000000000000000
  rate_multipliers:
    - 1000000000000000000
    - 1000000000000000000
  dy: 200000000000000000000
  calculated_dx: 200646751048726802272
- amp_coeff: 200
  initial_liquidity:
    - 1000000000000000000000
    - 1000000000000000000000
  rate_multipliers:
    - 1000000000000000000
    - 1000000000000000000
  dy: 200000000000000000000
  calculated_dx: 200813602707303095610
- amp_coeff: 200
  initial_liquidity:
    - 500000000000000000000
    - 10000000000000000000000
  rate_multipliers:
    - 1000000000000000000
    - 2000000000000000000
  dy: 200000000000000000000
  calculated_dx: 243982250129660465587
- amp_coeff: 1000
  initial_liquidity:
    - 500000000000000000000
    - 10000000000000000000000
  rate_multipliers:
    - 1000000000000000000
    - 2000000000000000000
  dy: 200000000000000000000
  calculated_dx: 359027584428509053001
- amp_coeff: 200
  initial_liquidity:
    - 1000000000000000000000
    - 1000000000000000000000
  rate_multipliers:
    - 1000000000000000000
    - 2000000000000000000
  dy: 200000000000000000000
  calculated_dx: 400329028520218852597
- amp_coeff: 1000
  initial_liquidity:
    - 1000000000000000000000
    - 1000000000000000000000
  rate_multipliers:
    - 1000000000000000000
    - 2000000000000000000
  dy: 200000000000000000000
  calculated_dx: 401047416416694999014
